go/vt/vtctl/grpcvtctldserver/server.go

```diff
@@ -430,9 +430,24 @@ func (s *VtctldServer) BackupShard(req *vtctldatapb.BackupShardRequest, stream v
 	span.Annotate("concurrency", req.Concurrency)
 	span.Annotate("incremental_from_pos", req.IncrementalFromPos)
 
-	tablets, stats, err := reparentutil.ShardReplicationStatuses(ctx, s.ts, s.tmc, req.Keyspace, req.Shard)
+	shardTablets, stats, err := reparentutil.ShardReplicationStatuses(ctx, s.ts, s.tmc, req.Keyspace, req.Shard)
+
+	var tablets []*topo.TabletInfo
+	// Instead of return on err directly, only return when no healthy tablets at all
 	if err != nil {
-		return err
+		for i, stat := range stats {
+			// shardTablets[i] and stats[i] is 1:1 mapping
+			// Healthy shardTablets[i] will be added to tablets
+			if stat != nil {
+				tablets = append(tablets, shardTablets[i])
+			}
+		}
+		// Only return err when all tablets have errors
+		if len(tablets) == 0 {
+			return err
```

Comment:
there is a small chance that we are doing `BackupShard --allow-primary` and the shard has exactly one tablet (the primary), and this could incorrectly fail. i'd suggest: `if len(tablets) < len(shardTablets)` instead

Reply:
Hi @ajm188 do you mean that the shard is built only with 1 tablet, so it's the only 1 and primary? 

if the code changed to:
```
if err != nil {
...
    if len(tablets) < len(shardTablets) {
        return err
    }
}
```
In normal cases (tablets >= 2), then it should equivelant to before and does not fix the bug., becase **len(tablets) < len(shardTablets)** should be always true when **err != nil**
```
if err != nil {
    return err
}
```

Instead, can I suggest tablets should always include the PRIMARY and no matter it's stats nil or non-nil?

```
shardTablets, stats, err := reparentutil.ShardReplicationStatuses(ctx, s.ts, s.tmc, req.Keyspace, req.Shard)

var tablets []*topo.TabletInfo
// Instead of return on err directly, only return when no healthy tablets at all
if err != nil {
	for i, stat := range stats {
		// Always include TabletType_PRIMARY
		if shardTablets[i].Type == topodatapb.TabletType_PRIMARY {
			tablets = append(tablets, shardTablets[i])
			continue
		}
		// shardTablets[i] and stats[i] is 1:1 mapping
		// Healthy shardTablets[i] will be added to tablets
		if stat != nil {
			tablets = append(tablets, shardTablets[i])
		}
	}
	// Only return err when no usable tablet
	if len(tablets) == 0 {
		return err
	}
} else {
	tablets = shardTablets
}
```

Reply:
ah good point.

> do you mean that the shard is built only with 1 tablet, so it's the only 1 and primary?

yes, that's what i mean.

i missed that the [`if len(tablets) == 0` happened inside the `if err != nil` check](https://github.com/vitessio/vitess/pull/14604/files#diff-4bb858fef99bc3779c1718d5dc3b3ef20d1efd16d80779d3e3318d9909f6749fR450-R455), not outside, so i think this is a moot point. sorry for the confusion!

Reply:
OK, given the intricate logic here I'd like to suggest extracting the logic into a separate function and adding a unit test suite to validate expected outcome of different scenarios 🙏

Reply:
I agree with Shlomi here

Reply:
updated PR